File: 01_Core_Platform/easycount-core/frontend/apps/admin-portal/src/api/accounting.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api } from "./client";

export interface TenantSettings {
  moneda: string;
  cuenta_ingresos: string | null;
  cuenta_itbis: string | null;
  cuenta_retenciones: string | null;
  dias_credito: number;
  correo_facturacion: string | null;
  telefono_contacto: string | null;
  notas: string | null;
  rounding_policy: string;
  odoo_sync_enabled: boolean;
  odoo_api_url: string | null;
  odoo_database: string | null;
  odoo_company_id: number | null;
  odoo_sales_journal_id: number | null;
  odoo_purchase_journal_id: number | null;
  odoo_fiscal_position_id: number | null;
  odoo_payment_term_id: number | null;
  odoo_currency_id: number | null;
  odoo_customer_document_type_id: number | null;
  odoo_vendor_document_type_id: number | null;
  odoo_credit_note_document_type_id: number | null;
  odoo_debit_note_document_type_id: number | null;
  odoo_sales_tax_id: number | null;
  odoo_purchase_tax_id: number | null;
  odoo_zero_tax_id: number | null;
  odoo_partner_vat_prefix: string | null;
  odoo_journal_code_hint: string | null;
  odoo_api_key_ref: string | null;
  updated_at: string;
}

export interface TenantSettingsUpdate {
  moneda: string;
  cuenta_ingresos: string | null;
  cuenta_itbis: string | null;
  cuenta_retenciones: string | null;
  dias_credito: number;
  correo_facturacion: string | null;
  telefono_contacto: string | null;
  notas: string | null;
  rounding_policy: string;
  odoo_sync_enabled: boolean;
  odoo_api_url: string | null;
  odoo_database: string | null;
  odoo_company_id: number | null;
  odoo_sales_journal_id: number | null;
  odoo_purchase_journal_id: number | null;
  odoo_fiscal_position_id: number | null;
  odoo_payment_term_id: number | null;
  odoo_currency_id: number | null;
  odoo_customer_document_type_id: number | null;
  odoo_vendor_document_type_id: number | null;
  odoo_credit_note_document_type_id: number | null;
  odoo_debit_note_document_type_id: number | null;
  odoo_sales_tax_id: number | null;
  odoo_purchase_tax_id: number | null;
  odoo_zero_tax_id: number | null;
  odoo_partner_vat_prefix: string | null;
  odoo_journal_code_hint: string | null;
  odoo_api_key_ref: string | null;
}

export interface LedgerTotals {
  total_emitidos: number;
  total_aceptados: number;
  total_rechazados: number;
  total_monto: string;
}

export interface LedgerStatusBreakdown {
  contabilizados: number;
  pendientes: number;
}

export interface LedgerMonthlyStat {
  periodo: string;
  cantidad: number;
  monto: string;
}

export interface AccountingSummary {
  totales: LedgerTotals;
  contabilidad: LedgerStatusBreakdown;
  series: LedgerMonthlyStat[];
}

export interface LedgerEntryItem {
  id: number;
  invoiceId?: number | null;
  invoice_id?: number | null;
  encf: string | null;
  referencia: string;
  cuenta: string;
  descripcion: string | null;
  debit: string;
  credit: string;
  fecha: string;
}

export interface LedgerPaginatedResponse {
  items: LedgerEntryItem[];
  total: number;
  page: number;
  size: number;
}

export interface LedgerEntryCreatePayload {
  invoice_id: number | null;
  referencia: string;
  cuenta: string;
  descripcion: string | null;
  debit: string;
  credit: string;
  fecha: string;
}

export function useTenantSettings(tenantId?: string) {
  return useQuery({
    queryKey: ["admin", "tenant-settings", tenantId],
    enabled: Boolean(tenantId),
    queryFn: async () => {
      const { data } = await api.get<TenantSettings>(`/api/v1/admin/tenants/${tenantId}/settings`);
      return data;
    },
  });
}

export function useUpdateTenantSettings(tenantId?: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (payload: TenantSettingsUpdate) => {
      const { data } = await api.put<TenantSettings>(`/api/v1/admin/tenants/${tenantId}/settings`, payload);
      return data;
    },
    onSuccess: async () => {
      await queryClient.invalidateQueries({ queryKey: ["admin", "tenant-settings", tenantId] });
    },
  });
}

export function useAccountingSummary(tenantId?: string) {
  return useQuery({
    queryKey: ["admin", "accounting-summary", tenantId],
    enabled: Boolean(tenantId),
    queryFn: async () => {
      const { data } = await api.get<AccountingSummary>(`/api/v1/admin/tenants/${tenantId}/accounting/summary`);
      return data;
    },
  });
}

export function useLedgerEntries(tenantId?: string, page = 1, size = 20) {
  return useQuery({
    queryKey: ["admin", "ledger-entries", tenantId, page, size],
    enabled: Boolean(tenantId),
    queryFn: async () => {
      const { data } = await api.get<LedgerPaginatedResponse>(`/api/v1/admin/tenants/${tenantId}/accounting/ledger`, {
        params: { page, size },
      });
      return data;
    },
  });
}

export function useCreateLedgerEntry(tenantId?: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (payload: LedgerEntryCreatePayload) => {
      const { data } = await api.post<LedgerEntryItem>(`/api/v1/admin/tenants/${tenantId}/accounting/ledger`, payload);
      return data;
    },
    onSuccess: async () => {
      await Promise.all([
        queryClient.invalidateQueries({ queryKey: ["admin", "ledger-entries", tenantId] }),
        queryClient.invalidateQueries({ queryKey: ["admin", "accounting-summary", tenantId] }),
      ]);
    },
  });
}

File: apps/easycount/app/models/accounting.py
"""Modelos para contabilidad y configuración de tenants."""
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Optional, TYPE_CHECKING

from sqlalchemy import Boolean, DateTime, ForeignKey, Numeric, String, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, utcnow
from app.models.tenant import Tenant

if TYPE_CHECKING:
    from app.models.invoice import Invoice


class TenantSettings(Base):
    """Parámetros contables y operativos por empresa."""

    __tablename__ = "tenant_settings"
    __table_args__ = (UniqueConstraint("tenant_id", name="uq_tenant_settings_tenant"),)

    tenant_id: Mapped[int] = mapped_column(ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    moneda: Mapped[str] = mapped_column(String(5), default="DOP")
    cuenta_ingresos: Mapped[Optional[str]] = mapped_column(String(64))
    cuenta_itbis: Mapped[Optional[str]] = mapped_column(String(64))
    cuenta_retenciones: Mapped[Optional[str]] = mapped_column(String(64))
    dias_credito: Mapped[int] = mapped_column(default=0)
    correo_facturacion: Mapped[Optional[str]] = mapped_column(String(255))
    telefono_contacto: Mapped[Optional[str]] = mapped_column(String(25))
    notas: Mapped[Optional[str]] = mapped_column(String(512))
    rounding_policy: Mapped[str] = mapped_column(String(32), default="HALF_UP")
    odoo_sync_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    odoo_api_url: Mapped[Optional[str]] = mapped_column(String(255))
    odoo_database: Mapped[Optional[str]] = mapped_column(String(120))
    odoo_company_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_sales_journal_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_purchase_journal_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_fiscal_position_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_payment_term_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_currency_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_customer_document_type_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_vendor_document_type_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_credit_note_document_type_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_debit_note_document_type_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_sales_tax_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_purchase_tax_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_zero_tax_id: Mapped[Optional[int]] = mapped_column(nullable=True)
    odoo_partner_vat_prefix: Mapped[Optional[str]] = mapped_column(String(12))
    odoo_journal_code_hint: Mapped[Optional[str]] = mapped_column(String(32))
    odoo_api_key_ref: Mapped[Optional[str]] = mapped_column(String(128))

    tenant: Mapped[Tenant] = relationship(backref="settings", single_parent=True, uselist=False)


class InvoiceLedgerEntry(Base):
    """Asiento contable simple asociado a un comprobante."""

    __tablename__ = "invoice_ledger_entries"

    tenant_id: Mapped[int] = mapped_column(ForeignKey("tenants.id", ondelete="CASCADE"), index=True)
    invoice_id: Mapped[int | None] = mapped_column(ForeignKey("invoices.id", ondelete="SET NULL"), nullable=True)
    referencia: Mapped[str] = mapped_column(String(64))
    cuenta: Mapped[str] = mapped_column(String(64))
    descripcion: Mapped[str | None] = mapped_column(String(255))
    debit: Mapped[Decimal] = mapped_column(Numeric(20, 6), default=Decimal("0"))
    credit: Mapped[Decimal] = mapped_column(Numeric(20, 6), default=Decimal("0"))
    fecha: Mapped[datetime] = mapped_column(DateTime, default=utcnow)

    tenant: Mapped[Tenant] = relationship(backref="ledger_entries")
    invoice: Mapped["Invoice | None"] = relationship(back_populates="ledger_entries")
